Add option to configure files with classpath: prefix or as base64 encoded strings

michaeljmarshall opened the pull request:
# Motivation

In certain environments, applications do not have predictable access to the file system for configuration, but do have access to configuration in memory or to the classpath. We can support these use cases by accepting the file content in the `PulsarConnectionFactory` configuration. This feature works out of the box for all configuration keys that end with `path`. The `PulsarConnectionFactory` will now take base64 encoded files and files with relative paths (based on the classpath).

Paths can now begin with `base64:` or with `classpath:` and will be correctly interpreted and passed to the Pulsar client. When the path's value is base64 encoded, the `PulsarConnectionFactory` will decode the file content, write the decoded file contents to a temporary file, and pass the temporary file path to the Pulsar Client and the Pulsar Admin Client. When the path's value is relative and starts with `classpath:`, the `PulsarConnectionFactory` will find the complete path and pass it to the Pulsar Client and the Pulsar Admin Client.

Note that the original config is not modified, so when the configuration is serialized, it retains the original values with `base64:` and `classpath:` prefixes.

# Rejected Alternative

We could have taken the file contents as an `InputStream` or as a `byte[]`. However, neither work well with the requirement that the `PulsarConnectionFactory` is serializable and deserializable.

michaeljmarshall commented:
@eolivelli - this is a draft so that I can add tests.

eolivelli commented on pulsar-jms/src/main/java/com/datastax/oss/pulsar/jms/Utils.java:
```diff
@@ -457,4 +462,45 @@ static int mapPriorityToPartitionNonLinearly(int priority, int numPartitions) {
+          }
+          Path file = Files.createTempFile("pulsar-jms.", ".tmp");
+          Files.setPosixFilePermissions(file, PosixFilePermissions.fromString("rw-------"));
+          file.toFile().deleteOnExit();
```
This is good
I wonder if we could return the list of these files and also eagerly delete them in Pulsar ConnectionFactory.clpse() (as well as keeping this deleteOnExit call)

michaeljmarshall replied:
Yes, that definitely makes sense.

michaeljmarshall commented:
@eolivelli - great point about the serde tests. I discovered that `byte[]` won't work. Since it wasn't requested by the customer, I removed it. I also realized that it is trivial to add support for `classpath:` prefixes, and since this was explicitly requested, I added it.

lhotari commented on pulsar-jms/src/main/java/com/datastax/oss/pulsar/jms/Utils.java:
```diff
@@ -457,4 +464,53 @@ static int mapPriorityToPartitionNonLinearly(int priority, int numPartitions) {
+                    "Cannot decode base64 " + key + " and create temporary file: " + ex, ex);
+          }
+        } else if (originalPath.startsWith("classpath:")) {
+          URL path = Utils.class.getClassLoader().getResource(originalPath.substring("classpath:".length()));
+          if (path == null) {
+            throw new IllegalArgumentException("File " + originalPath + " not found for " + key);
+          }
+          finalPath = path.getPath();
+          log.info("Resolved {} to {}", key, finalPath);
```
This wouldn't work for all cases. For example, for a resource inside a jar file. For a generic solution, it would be necessary to copy to a temp file.

eolivelli replied:
I agree, we must copy to a temporary file, because with this way it works only if you have a directory (like an exploded jar) in the classpath.

I believe that it is always better to download the resource and put it into a temporary file

michaeljmarshall replied:
Thanks for your feedback. I didn't realize that was necessary. I'll fix it now.

lhotari commented on pulsar-jms/src/test/resources/README.md:
```diff
@@ -0,0 +1,24 @@
+
+openssl genrsa -out ca.key 2048
+
+openssl req -x509 -subj '/CN=localhost' -new -nodes -key ca.key \
```
It's not really an issue in this PR, since this is a problem in most Pulsar TLS instructions that SAN isn't used for certificates. I've always wondered about the mix of using openssl and keytool. keytool alone could create proper certificates. With openssl, adding a SAN is possible, but it's a bit awkward. Here's an example from a script for renewing Pulsar test certs:
https://github.com/apache/pulsar/blob/6cc30c93a05bb16b7ef62ffebc15ceb5e6efb502/build/regenerate_certs_for_tests.sh#L37

My assumption is that using keytool alone would be a more portable way.
By googling, I found this example: https://ultimatesecurity.pro/post/san-certificate/ 
On Stackoverflow, https://stackoverflow.com/questions/8744607/how-to-add-subject-alernative-name-to-ssl-certs

eolivelli commented on pulsar-jms/src/main/java/com/datastax/oss/pulsar/jms/Utils.java:
```diff
@@ -457,4 +464,53 @@ static int mapPriorityToPartitionNonLinearly(int priority, int numPartitions) {
+                    "Cannot decode base64 " + key + " and create temporary file: " + ex, ex);
+          }
+        } else if (originalPath.startsWith("classpath:")) {
+          URL path = Utils.class.getClassLoader().getResource(originalPath.substring("classpath:".length()));
```
Using Thread.currentThread().getContextClassload() is generally better, because otherwise you are using only the classloader that loaded the "Utils" class, that is in the pulsar-jms-all jar, and it is not ensured that you are able to access the other resources deployed with the application.

Also usually in JavaEE environments the TCCL (Thread Context Class Loader) is the one you use to load "contextual" resources (Classes with reflection, resource files...)

michaeljmarshall replied:
Makes sense, great catch.

eolivelli reviewed (changes requested):
Great work

I have left a couple of comments about "classpath:"

michaeljmarshall commented:
Closing this PR. I am going to port this upstream to contribute it to the Apache Pulsar Java and Admin Clients.
